File: 2017/12.py
import heapq

connections = {}
with open("input_12.txt") as f:
    for l in f:
        if l.startswith("###"):
            break
        l = l.split("\n")[0].split("<->")
        connections[int(l[0])] = list(map(int, l[1].split(",")))

all_progs = sorted(connections.keys())
groups = []
while len(all_progs) != 0:
    progs = []
    heapq.heappush(progs, all_progs[0])
    group = set()
    while len(progs) != 0:
        p = heapq.heappop(progs)
        for candidate in connections[p]:
            if candidate not in group:
                heapq.heappush(progs, candidate)
        group.update([p])

    groups.append((group, len(group)))
    for g in group:
        all_progs.remove(g)
print ("🎄Part 1: {}".format(groups[0][1]))
print ("🎁🎄Part 2: {}".format(len(groups))) 

File: 2019/12.jl
include("utils.jl")
using .AoCUtils
using DataStructures

function run()
    connections = Dict()
    lines = AoCUtils.readLines("input_12.txt")
    for l in lines
        l = split(l, "<->")
        connections[parse(Int, l[1])] = parse.(Int, split(l[2], ","))
    end

    all_progs = sort(collect(keys(connections)))
    groups = []
    while length(all_progs) != 0
        progs = MutableBinaryMinHeap{Int}([])
        push!(progs, all_progs[1])
        group = Set()
        while length(progs) != 0
            p = pop!(progs)
            for candidate in connections[p]
                if !(candidate in group)
                    push!(progs, candidate)
                end
            end
            push!(group, p)
        end
        push!(groups, (group, length(group)))
        for g in group
            filter!(x->x!=g, all_progs)
        end
    end
    println("🎄 $(groups[1][2])")
    println("🎁🎄 $(length(groups))") 
end

run()
